cachito/workers/pkg_managers/pip.py

```diff
@@ -1203,3 +1208,264 @@ def finalize_nexus_for_pip_request(pip_repo_name, raw_repo_name, username):
+    :param bool require_hashes: True if all requirements must specify a checksum
+    :raise ValidationError: If any requirement does not meet expectations
+    """
+    # Fail if any PyPI dependency is not pinned to an exact version
+    for req in filter(lambda r: r.kind == "pypi", requirements):
+        vspec = req.version_specs
+        if len(vspec) != 1 or vspec[0][0] not in ("==", "==="):
+            msg = f"Requirement must be pinned to an exact version: {req.download_line}"
+            raise ValidationError(msg)
+
+    # Fail if any dependency requires a hash but does not specify one
+    for req in requirements:
+        hash_required = require_hashes or req.kind == "url"
+
+        if hash_required and not req.hashes and not req.qualifiers.get("cachito_hash"):
+            msg = f"Hash is required, dependency does not specify any: {req.download_line}"
+            raise ValidationError(msg)
+
+
+def _download_pypi_package(requirement, pip_deps_dir, pypi_url, pypi_auth):
+    """
+    Download the sdist (source distribution) of a PyPI package.
+
+    The package must be pinned to an exact version using the '==' (or '===') operator.
+    While the specification defines the '==' operator as slightly magical (reference:
+    https://www.python.org/dev/peps/pep-0440/#version-matching), we treat the version
+    as exact.
+
+    Does not download any dependencies (implied: ignores extras). Ignores environment
+    markers (target environment is not known to Cachito).
+
+    :param PipRequirement requirement: PyPI requirement from a requirement.txt file
+    :param Path pip_deps_dir: The deps/pip directory in a Cachito request bundle
+    :param str pypi_url: URL of PyPI (proxy) server
+    :param requests.auth.AuthBase pypi_auth: Authorization for the PyPI server
+
+    :return: Path to downloaded file
+    """
+    package = requirement.package
+    version = requirement.version_specs[0][1]
+
+    # See https://warehouse.readthedocs.io/api-reference/json/
+    package_url = f"{pypi_url.rstrip('/')}/pypi/{package}/{version}/json"
+    try:
+        pypi_resp = requests_session.get(package_url, auth=pypi_auth)
+        pypi_resp.raise_for_status()
+    except requests.RequestException as e:
+        raise CachitoError(f"PyPI query failed: {e}")
+
+    data = pypi_resp.json()
+    sdists = [pkg for pkg in data["urls"] if pkg["packagetype"] == "sdist"]
+    if not sdists:
+        raise CachitoError(f"No sdists found for package {package}=={version}")
+
+    # Choose best candidate based on sorting key
+    sdist = max(sdists, key=_sdist_preference)
+    if sdist["yanked"]:
+        raise CachitoError(f"All sdists for package {package}=={version} are yanked")
+
+    try:
+        file_resp = requests_session.get(sdist["url"], stream=True)
+        file_resp.raise_for_status()
+    except requests.RequestException as e:
+        raise CachitoError(f"Could not download {sdist['filename']}: {e}")
+
+    package_dir = pip_deps_dir / package
```

Comment:
Just wondering about this syntax out of curiosity. What do these extra spaces around `/` indicate if we are defining a simple path?

Reply:
@nirzari, the slash operator is special for pathlib path objects. It works like os.path.join.

See https://docs.python.org/3/library/pathlib.html#operators